pkg/domain/infra/abi/generate.go

```diff
@@ -41,6 +45,81 @@ func (ic *ContainerEngine) GenerateSystemd(ctx context.Context, nameOrID string,
 	return &entities.GenerateSystemdReport{Units: units}, nil
 }
 
+func (ic *ContainerEngine) GenerateSpec(ctx context.Context, opts *entities.GenerateSpecOptions) (*entities.GenerateSpecReport, error) {
+	var spec *specgen.SpecGenerator
+	var pspec *specgen.PodSpecGenerator
+	var err error
+	if _, err := ic.Libpod.LookupContainer(opts.ID); err == nil {
+		spec = &specgen.SpecGenerator{}
+		_, _, err = generateUtils.ConfigToSpec(ic.Libpod, spec, opts.ID)
+		if err != nil {
+			return nil, err
+		}
+	} else if p, err := ic.Libpod.LookupPod(opts.ID); err == nil {
+		pspec = &specgen.PodSpecGenerator{}
+		pspec.Name = p.Name()
+		_, err := generateUtils.PodConfigToSpec(ic.Libpod, pspec, &entities.ContainerCreateOptions{}, opts.ID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if pspec == nil && spec == nil {
+		return nil, fmt.Errorf("could not find a pod or container with the id %s", opts.ID)
+	}
+
+	// rename if we are looking to consume the output and make a new entity
+	if opts.Clone {
+		if spec != nil {
+			spec.Name = generateUtils.CheckName(ic.Libpod, spec.Name, "ctr")
+		} else {
+			pspec.Name = generateUtils.CheckName(ic.Libpod, pspec.Name, "pod")
+		}
+	}
+
+	j := []byte{}
+	if spec != nil {
+		j, err = json.MarshalIndent(spec, "", " ")
+		if err != nil {
+			return nil, err
+		}
+	} else if pspec != nil {
+		j, err = json.MarshalIndent(pspec, "", " ")
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if !opts.Compact && len(opts.FileName) > 0 {
+		err = ioutil.WriteFile(opts.FileName, j, 0644)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if opts.Verbose {
+		// compact output
+		if opts.Compact {
+			compacted := &bytes.Buffer{}
+			err := json.Compact(compacted, j)
+			if err != nil {
+				return nil, err
+			}
+			fmt.Println(compacted.String())
+			if len(opts.FileName) > 0 {
+				// put in a file if requested
+				err = ioutil.WriteFile(opts.FileName, compacted.Bytes(), 0644)
+				if err != nil {
+					return nil, err
+				}
+			}
+			// regular output
+		} else {
+			fmt.Println(string(j))
+		}
+	}
```

Comment:
Please do not use prints under pkg/..., output should be only handled in cmd/podman... for a consistent structure.
I would just return the byte array and then the cmd/podman can either write to a file or stdout.